docs/RationaleHW.md

```diff
@@ -451,7 +455,90 @@ All that said, using attributes is the right thing for a number of reasons:
+within them.
+
+*Type Declaration Operation*
+
+A [Type Declaration](https://circt.llvm.org/docs/Dialects/HW/#hwtypedecl-circthwtypedeclop)
+declares a symbolic name for a type. It consists of:
+
+* A symbolic name, which can be referred to in the IR by a Type Alias Type.
+* A type attribute, which contains the underlying type the name refers to
+* An optional string attribute, which specifies a name to give the type in the
+  output. If not specified, the symbolic name is used in the output.
+
+*Type Alias Type*
+
+A [Type Alias](https://circt.llvm.org/docs/Dialects/HW/#an-symbolic-reference-to-a-type-declaration)
+refers to a type declaration symbolically. It consists of:
+
+* A symbolic reference to a Type Scope and Type Declaration within the scope
+* A copy of the underlying type, cached in type storage
+* A copy of the canonical type, cached in type storage
+
+**Type declaration System Verilog output**
+
+In ExportVerilog, Type Scopes may be split into their own output file, included
+at the top of multiple split output files, or included via a header file.
+
+Each Type Declaration will be emitted according to the System Verilog spec,
+section 6.18, User-defined types. For example:
+
+```
+typedef logic mytype;
+```
+
+**Type canonicalization**
+
+We take an approach to type canonicalization similar to [Clang](https://clang.llvm.org/docs/InternalsManual.html#canonical-types).
+
+To implement this, CIRCT has an extra layer of indirection in the ODS
+declarations of types that may be targetted by Type Declarations. This layer
+makes it possible to generically let operations declare their operands and
+results to be of a specific type, and also accept type declarations that
+canonicalize to that type.
+
+The HW dialect types are declared in [HWTypesImpl.td](https://github.com/llvm/circt/blob/main/include/circt/Dialect/HW/HWTypesImpl.td),
+which declares the actual type and is used to generate the C++ wrappers. To make
+these types usable in ODS operations in a generic way that support type
+declarations, wrapper types are declared in [HWTypes.td](https://github.com/llvm/circt/blob/main/include/circt/Dialect/HW/HWTypes.td),
+which uses helpers to define dialect types that may be either the declared type,
+or a type declaration whose canonical type is the declared type.
+
+In order to "see through" type declarations and get at the canonical type, a set
+of helper functions are added to the hardware dialect: `hw::type_isa`,
+`hw::type_cast`, and `hw::type_dyn_cast`. These should generally be used instead
+of the MLIR `Type::isa`, `Type::cast`, and `Type::dyn_cast`, because they
+transparently support type declarations.
+
+As the Clang docs state: "The only hard part here is remembering not to use the
+isa/cast/dyn_cast operations."
+
+**Open Issues**
+
+*Duplicated type in Type Alias Type IR*
+
+In order to support Type canonicalization, the Type Alias Type keeps a copy of
```

Comment:
Not super familiar with this, but I'm curious. Is it possible for canonicalization to apply multiple times while converging, e.g., A -> B -> C? If so, is it possible to refer back to A, given C?